engine/table/src/main/java/io/deephaven/engine/table/impl/remote/ConstructSnapshot.java

```diff
@@ -1509,61 +1531,184 @@ private static boolean serializeAllTable(
                     .append(", rows=").append(snapshot.rowsIncluded).append("/").append(keysToSnapshot)
                     .append(", cols=");
-            if (columnsToSerialize == null) {
+            if (columnsToSnapshot == null) {
                 logEntry.append("ALL");
             } else {
-                logEntry.append(FormatBitSet.formatBitSet(columnsToSerialize));
+                logEntry.append(FormatBitSet.formatBitSet(columnsToSnapshot));
             }
             logEntry.append(", usePrev=").append(usePrev).endl();
         }
+        return true;
+    }
 
+    /**
+     * Check if all the required column sources are in memory and should allow efficient access.
+     */
+    private static boolean allColumnSourcesInMemory(
+            @NotNull final Table table,
+            @NotNull final String[] columnSources,
+            @NotNull final TIntList columnIndicesToSnapshot) {
+        final int numColumnsToSnapshot = columnIndicesToSnapshot.size();
+        for (int colRank = 0; colRank < numColumnsToSnapshot; ++colRank) {
+            final ColumnSource<?> columnSource =
+                    table.getColumnSource(columnSources[columnIndicesToSnapshot.get(colRank)]);
+            if (!isColumnSourceInMemory(columnSource)) {
+                return false;
+            }
+        }
         return true;
     }
 
-    private static boolean serializeAllTables(
+    /**
+     * Recursively check if the column source is in-memory or redirected to an in-memory column source.
+     */
+    private static boolean isColumnSourceInMemory(final ColumnSource<?> columnSource) {
+        if (columnSource instanceof InMemoryColumnSource) {
+            return true;
+        }
+        if (!(columnSource instanceof RedirectedColumnSource)) {
+            return false;
+        }
+        final ColumnSource<?> innerSource = ((RedirectedColumnSource<?>) columnSource).getInnerSource();
+        return isColumnSourceInMemory(innerSource);
+    }
+
+    private static boolean snapshotColumnsParallel(
+            final String[] columnSources,
+            @NotNull final TIntList columnIndicesToSnapshot,
+            @NotNull final Table table,
+            final boolean usePrev,
+            final ExecutionContext executionContext,
+            @NotNull final BarrageMessage snapshot) {
+        final JobScheduler jobScheduler = new OperationInitializerJobScheduler();
+        final CompletableFuture<Void> waitForParallelSnapshot = new CompletableFuture<>();
+        jobScheduler.iterateParallel(
+                executionContext,
+                logOutput -> logOutput.append("snapshotColumnsParallel"),
+                JobScheduler.DEFAULT_CONTEXT_FACTORY,
+                0, columnIndicesToSnapshot.size(),
+                (context, colRank, nestedErrorConsumer) -> snapshotColumnsSerial(
+                        columnSources,
+                        new TIntArrayList(new int[] {columnIndicesToSnapshot.get(colRank)}),
```

Comment:
Ok, starting to see why you preferred the int list.

Another alternative would be to use a RowSet to indicate which columns to use. Single key RowSets are probably more friendly. I would discuss with Ryan whether to use BitSet vs RowSet vs TIntArrayList. (I'm leaning toward RowSet now.)

Reply:
Will check with Ryan.